miner/bidder.go

```diff
@@ -0,0 +1,232 @@
+package miner
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var (
+	dialer = &net.Dialer{
+		Timeout:   time.Second,
+		KeepAlive: 60 * time.Second,
+	}
+
+	transport = &http.Transport{
+		DialContext:         dialer.DialContext,
+		MaxIdleConnsPerHost: 50,
+		MaxConnsPerHost:     50,
+		IdleConnTimeout:     90 * time.Second,
+		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true},
+	}
+
+	client = &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: transport,
+	}
+)
+
+type ValidatorConfig struct {
+	Address common.Address
+	URL     string
+}
+
+type BidderConfig struct {
+	Enable     bool
+	Validators []ValidatorConfig
+}
+
+type Bidder struct {
+	config *BidderConfig
+	engine consensus.Engine
+	chain  *core.BlockChain
+
+	validators map[common.Address]*ethclient.Client // validator address -> ethclient.Client
+
+	bestWorksMu sync.RWMutex
+	bestWorks   map[int64]*environment
+}
+
+func NewBidder(config *BidderConfig, engine consensus.Engine, chain *core.BlockChain) *Bidder {
+	b := &Bidder{
+		config:     config,
+		engine:     engine,
+		chain:      chain,
+		validators: make(map[common.Address]*ethclient.Client),
+		bestWorks:  make(map[int64]*environment),
+	}
+
+	if !config.Enable {
+		return b
+	}
+
+	for _, v := range config.Validators {
+		cl, err := ethclient.DialOptions(context.Background(), v.URL, rpc.WithHTTPClient(client))
+		if err != nil {
+			log.Error("Bidder: failed to dial validator", "url", v.URL, "err", err)
+			continue
+		}
+
+		b.validators[v.Address] = cl
+	}
+
+	if len(b.validators) == 0 {
+		log.Warn("Bidder: No valid validators")
+	}
+
+	return b
+}
+
+// Bid called by go routine
+// 1. ignore and return if bidder is disabled
+// 2. panic if no valid validators
+// 3. ignore and return if the work is not better than the current best work
+// 4. update the bestWork and do bid
+func (b *Bidder) Bid(work *environment) {
+	if !b.config.Enable {
+		log.Warn("Bidder: disabled")
+		return
+	}
+
+	if len(b.validators) == 0 {
+		log.Crit("Bidder: No valid validators")
+		return
+	}
+
+	// if the work is not better than the current best work, ignore it
+	if !b.isBestWork(work) {
```

Comment:
The validator only accept 3 bids from a builder at most, so I think we should not send every bid

Reply:
> The validator only accept 3 bids from a builder at most, so I think we should not send every bid.
> 
> So I think it is better there is a independent routine, it will pick the best work 3 times during a blocking time and send to validators.

This is important, please resolve.

Reply:
I check the latest code 
```
		case <-timer.C:
			works := b.works[currentHeight]
			slices.SortStableFunc(works, func(i, j *environment) int {
				return j.profit.Cmp(i.profit)
			})

			for i, work := range works {
				// only bid for the top 3 most profitable works
				if i >= 3 {
					break
				}
				b.bid(work)
			}
```

what we expect is that a builder send the top 3 bid at different time considerring the network delay and builder compitition, but not send the three at the same time (if you already know which one is the best, why bother send three?)